feat: add SES experiment toggle (iOS)

leotm opened the pull request:
## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

A feature toggle for SES, within our settings menu under experiments, enabled by default

TODO
- [x] add `react-native-mmkv`
- [x] MVP
- [x] ensure only visible on iOS not Android
  - show only on iOS
  - or show if storage key exists
- [x] check next to Blockaid experiment enabled
- [x] fix test, update snapshot
- [x] add descriptive content
- [x] translate text
- [x] update styling
- [x] <s>replace `react-native-mmkv` with `react-native-default-preference`</s>
  - attempted in https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f but only [partially working](https://github.com/MetaMask/metamask-mobile/pull/8373#issuecomment-1908544768)
- [x] fix Android debug bundling on Hermes
- [x] <s>consider content revision to include _Hardened JS_  and _SES (Secure EcmaScript)_ terms</s>
- [x] update to newly discussed text with link
- [x] fix styling, move switch above next to title
- [x] display both SES and Blockaid under same _Security_ heading
- [x] preserve Android UI
- [x] update SES to be enabled by default

## **Related issues**

Fixes: partially mitigates rollout risk

Follow-up to
- https://github.com/MetaMask/metamask-mobile/pull/8033

## **Manual testing steps**

iOS
- boot app
- login / create wallet
- go to bottom tab Settings menu
- open Experimental menu
- _SES enabled by default, like currently in (main) branch_
- disable SES ft toggle
- reboot app
- _SES now disabled_
- verify in Experimental menu
- repeat steps above to re-enable

Android
- boot app
- login / create wallet
- go to bottom tab Settings menu
- open Experimental menu
- UI remains the same
  - Blockaid shown if feature is enabled

## **Screenshots/Recordings**

Previous MVP video (old UI) [demo](https://consensys.slack.com/archives/C02U025CVU4/p1706035121992709?thread_ts=1705527596.738189&cid=C02U025CVU4) showing the behaviour, a simple settings menu ft toggle that persists, enabling/disabling lockdown upon app reboot, on iOS only

Android remains unchanged ([Blockaid disabled](https://github.com/MetaMask/metamask-mobile/assets/1881059/d74a30be-e65e-4832-a21e-8aa1812df608), [Blockaid enabled](https://github.com/MetaMask/metamask-mobile/assets/1881059/77b9a24b-486b-4e6b-aa10-51eede8c2951))

### **Before**

Blockaid ft disabled

<img width="403" alt="Screenshot 2024-01-26 at 7 20 46 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/ec120e0a-9ba5-454c-b4c8-21307ced7ea5">

Blockaid ft enabled

<img width="402" alt="Screenshot 2024-01-26 at 7 25 05 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/4be5fda0-edfe-427f-8206-df6df100a6fc">

### **After**

Blockaid ft disabled

<img width="403" alt="Screenshot 2024-01-30 at 1 07 07 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/bca71f83-ab01-4d70-bbc1-fac00dd76e23">

Blockaid ft enabled

<img width="401" alt="Screenshot 2024-01-30 at 1 07 48 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/7d12aa9a-51e9-44d8-a6d0-6d9ca8aa9375">

Navigating to [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md)

<img width="401" alt="Screenshot 2024-01-26 at 6 52 17 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/a32d47b3-e95e-4a15-b745-3cbc7c444250">

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

leotm committed:
Add react-native-mmkv

- add NPM module
- update Yarn lockfile
- update CocoaPods lockfile

leotm committed:
MVP

leotm commented on patches/react-native+0.71.15.patch:
```diff
@@ -1,16 +1,18 @@
  'use strict';
  
-+const Platform = require('../Utilities/Platform');
++import { MMKV } from 'react-native-mmkv';
```
no luck with `require` here

<img width="415" alt="Screenshot 2024-01-23 at 5 15 58 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/1edaa459-a6d0-4770-922d-c5e1d10baf72">
<img width="404" alt="Screenshot 2024-01-23 at 5 16 07 pm" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/8f67a3d5-b5e2-47cf-8cfa-820531d1f696">

 but `import` works

leotm replied:
mmkv import breaking Hermes debug bundle on Android archs
splitting toggle logic to file required on iOS only should be sufficient
due to mmkv's async functions/generators unsupported on Hermes

leotm committed:
Only show on iOS

sethkfman commented:
@gauthierpetetin @Cal-L Can you take a look at this contribution to make sure it aligns with our platform for a product perspective?

@leotm Is there a more descriptive Header and content we can add so users and understand? Maybe something like Lockdown - This feature prevents dynamic modification of JavaScript running in the application.

gauthierpetetin commented:
Hi @sethkfman , sorry I only see this now. I think end users won't understand the feature in its current form. I'll setup a meeting with @yanrong-chen @coreyjanssen @leotm and @hesterbruikman to discuss how we could present it.

leotm committed:
Update snapshot

leotm committed:
Remove stale comment

leotm committed:
Update styling

leotm committed:
Update header

leotm committed:
Update content

leotm committed:
Translate header

leotm committed:
Translate description

leotm committed:
Update snapshot

leotm committed:
Replace RN MMKV with RN default-preference

- refactor to callbacks
- refactor booleans to strings
- refactor conditions
- DefaultPreference working in settings menu only

Blocker: DefaultPreference.get value remains undefined called at Initializecore

leotm committed:
Revert "Replace RN MMKV with RN default-preference"

This reverts commit ea5674c5600d5d11d78a07ff4ab774cf3848989f.

leotm commented:
@sethkfman i refactored to replace mmkv with default-preference https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f
but had to revert to mmkv, since default-preference only partially works here as a toggle
(fine in settings menu, but in InitializeCore [.get](https://github.com/MetaMask/metamask-mobile/pull/8373/commits/ea5674c5600d5d11d78a07ff4ab774cf3848989f#diff-982477f8c8bf57eb44d53f104ddf22fb69df167560a37ad981551317419bb46bR13) is unable to return the storage value)

jpcloureiro commented on app/components/Views/Settings/ExperimentalSettings/index.tsx:
```diff
@@ -164,9 +176,47 @@ const ExperimentalSettings = ({ navigation, route }: Props) => {
   return (
     <ScrollView style={styles.wrapper}>
       <WalletConnectSettings />
+      {Platform.OS === 'ios' && <SesSettings />}
```
We can use `isIos` boolean from `./app/util/device` instead

https://github.com/MetaMask/metamask-mobile/blob/2b66c53abbcf63440c179e8bebf64b0033c804d1/app/util/device/index.js#L15-L17

leotm replied:
updated ^^

leotm committed:
Use Device.isIos fn

leotm committed:
Update import

leotm committed:
Remove stale import

leotm committed:
Fix Android (Hermes) debug bundle

Resolve: https://github.com/MetaMask/metamask-mobile/pull/8373#discussion_r1466388666

leotm committed:
Add MetaMask copyright notice

leotm committed:
Document SES disabled in debug-mode

leotm committed:
Merge branch 'main' into feat/ses-experimental-setting

leotm committed:
Run Yarn deduplicate

leotm committed:
Add back missing import

leotm committed:
Merge branch 'main' into feat/ses-experimental-setting

leotm committed:
Update snapshot

gauthierpetetin commented:
Suggestion agreed with @coreyjanssen @leotm and @hesterbruikman :

- Title: Code lockdown
- Description: This is a test feature that blocks any changes to the app's javascript code without permission. [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md).

leotm committed:
Update heading

leotm committed:
Add SES URL

leotm committed:
Fix SES URL

- #readme doesn't scroll down on iOS
- update to actual README.md file

leotm committed:
Update description

leotm committed:
Update description

leotm committed:
Update UI

leotm committed:
Update UI

leotm committed:
Preserve UI on Android

hesterbruikman commented:
@naugtur IMO we should not release SES without a toggle to disable. My understanding is that conventional dapp use will not break, but we can't possibly test for all dapp use cases and so users need to be able to turn this feature off.

Is there a reason this PR is in draft?

If SES is already in 7.16.0 RC I imagine we want to add merge and add this PR asap.

- Has any `Learn more` copy been prepared? @naugtur please let me know so we can involve the documentation team to add knowledge base content to link out to.
- Believe this needs a translation PR for the copy

leotm commented:
> @naugtur IMO we should not release SES without a toggle to disable. My understanding is that conventional dapp use will not break, but we can't possibly test for all dapp use cases and so users need to be able to turn this feature off.
> 
> Is there a reason this PR is in draft?
> 
> If SES is already in 7.16.0 RC I imagine we want to add merge and add this PR asap.
> 
> * Has any `Learn more` copy been prepared? @naugtur please let me know so we can involve the documentation team to add knowledge base content to link out to.
> * Believe this needs a translation PR for the copy

agreed been pushing to get this merged last week prior our Thurs RC cut (extended to Fri)

so presuming this will need cherry-pick into 7.16.0

i converted this PR to draft only now, only to make the change of SES enabled by default instead of disabled
then updating the screenshots and re-readying up
(i think this is more convenient than merging this, then 2 PRs to be cherry-picked)

> Suggestion agreed with @coreyjanssen @leotm and @hesterbruikman :
> 
> * Title: Code lockdown
> * Description: This is a test feature that blocks any changes to the app's javascript code without permission. [Learn more](https://github.com/endojs/endo/blob/master/packages/ses/README.md).

last Fri we agreed the Learn more link to be this above, who should i speak to clarify this is ok?

leotm committed:
Update React Native core SES comment

leotm committed:
Enable SES by default

hesterbruikman commented:
Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?

I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.

leotm commented:
> Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?

np! if by feature we mean this toggle for SES, the trade-off: our users might disable SES (reduce security), in the hopes of fixing an edge case of functionality (not yet covered by our tests)

or if by feature we mean SES, the trade-off: better security, at the compromise of possibly encountering an edge case of functionality not working as expected (i.e. discovery of a library we're using that calls non-standard JS)

> I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.

exactly the toggle is a temp Experimental fallback, long-term we want SES enabled indefinitely (like in mm-extension) for better security (permanently prevent prototype pollution - now that's a tongue twister)

sethkfman reviewed (dismissed):
LGTM

sethkfman committed:
Merge branch 'main' into feat/ses-experimental-setting

hesterbruikman commented:
> > Thanks @leotm ! And apologies, hadn't noticed the link for Learn more. Can you please explain the trade-off in enabling this feature?
> 
> np! if by feature we mean this toggle for SES, the trade-off: our users might disable SES (reduce security), in the hopes of fixing an edge case of functionality (not yet covered by our tests)
> 
> or if by feature we mean SES, the trade-off: better security, at the compromise of possibly encountering an edge case of functionality not working as expected (i.e. discovery of a library we're using that calls non-standard JS)
> 
> > I'd like to understand if we can treat the toggle as a temporary Experimental fallback, in case of unforeseen issues. Or if long-term there might always be end-user use cases to disable the feature under Advanced. In case of the latter we can reach out to the documentation team to create Knowledge Base content.
> 
> exactly the toggle is a temp Experimental fallback, long-term we want SES enabled indefinitely (like in mm-extension) for better security (permanently prevent prototype pollution - now that's a tongue twister), then ofc run LavaMoat (which [runs](https://github.com/LavaMoat/LavaMoat/blob/main/packages/core/src/kernelTemplate.js#L54-L67) on SES)

Just to conclude on this: I don't think we need further `Learn more` end-user documentation.

sethkfman committed:
fix patch issue by moving the error logging to the new portion of patch
